internal/server/data/data.go

```diff
@@ -152,28 +139,62 @@ func add[T models.Modelable](db *gorm.DB, model *T) error {
 		return err
 	}
 
-	if err := db.Create(model).Error; err != nil {
-		if isUniqueConstraintViolation(err) {
-			return fmt.Errorf("%w: %s", internal.ErrDuplicate, err)
-		}
+	err := db.Create(model).Error
+	return handleError(err)
+}
 
-		return err
-	}
+type UniqueConstraintError struct {
+	Table  string
+	Column string
+}
 
-	return nil
+func (e UniqueConstraintError) Error() string {
+	return fmt.Sprintf("value for %v already exists in %v", e.Column, e.Table)
 }
 
-func isUniqueConstraintViolation(err error) bool {
-	var pgerr *pgconn.PgError
-	if errors.As(err, &pgerr) {
-		return pgerr.Code == pgerrcode.UniqueViolation
+// handleError looks for well known DB errors. If the error is recognized it
+// is translated into a UniqueConstraintError so that calling code can
+// inspect the error.
+func handleError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) {
+		switch pgErr.Code {
+		case pgerrcode.UniqueViolation:
+			// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/access/nbtree/nbtinsert.c;h=f6f4af8bfe3f0876e860944e2233328f8cc3303a;hb=HEAD#l668
+			// ereport(ERROR,
+			//         (errcode(ERRCODE_UNIQUE_VIOLATION),
+			//          errmsg("duplicate key value violates unique constraint \"%s\"",
+			//                 RelationGetRelationName(rel)),
+			//                 key_desc ? errdetail("Key %s already exists.",
+			//                            key_desc) : 0,
+			//                 errtableconstraint(heapRel,
+			//                                    RelationGetRelationName(rel))));
+			re := regexp.MustCompile(`Key \((?P<columnName>\w+)\)=\(\w+\) already exists.`)
+			matches := re.FindStringSubmatch(pgErr.Detail)
```

Comment:
Aren't the fields already available from `pgErr` ? I thought we could use `pgErr.TableName`, and `pgErr.ColumnName`.